### PR TITLE
OS#16317718: Set isNotPathTypeHandlerOrHasUserDefinedCtor flag for Ctors as early as possible

### PR DESCRIPTION
**Pull request by irinayat-MS**
Fixes https://microsoft.visualstudio.com/OS/_workitems?id=16317718

Change 2f8107549 introduced optimization of path handlers that loses "custom ctor" flag because the new type in SetConfigurable is branched off a parent and not the type created when the property is added. 

Does it make sense to move handling of the flag from AddPropertyInternal into PromoteType<T>?

All create-ctor-poisoned.js tests from test262\built-ins\Array pass with this change.

### CODE REVIEW

**Comment by boingoing** [comment]
We should add some unit test coverage for this - test262 is not run in the CI.

**Comment by pleath** [comment]
Yes, it seems necessary to move the setting of that flag into PromoteType, now that PromoteType is called from various places to build up a type path.

File: lib/Runtime/Types/PathTypeHandler.cpp [hunk 1]
```diff
@@ -2096,6 +2096,11 @@ namespace Js
 
         (*propertyIndex) = index;
 
+        if (key.GetPropertyId() == PropertyIds::constructor)
```
Comment (pleath) [inline lib/Runtime/Types/PathTypeHandler.cpp]: Can this code be moved to the point in PromoteType where we do install a new type handler, say at line 2035? On the other path we're not creating a new type handler, so the field is already in the right state. Thanks.

**Comment by irinayat-MS** [comment]
Updated map/filter tests (concat, slice and splice also use ArraySpeciesCreate but for this particular bug any of the built-in methods provides sufficient coverage)

---
In reply to: [372851012](https://github.com/Microsoft/ChakraCore/pull/4814#issuecomment-372851012) [](ancestors = 372851012)

**Pull request merged**